ocaml/xapi/stream_vdi.ml

```diff
@@ -113,53 +208,95 @@ let send_all refresh_session ofd ~__context rpc session_id (prefix_vdis: vdi lis
+                write_block ~__context filename Bytes.empty ofd 0;
+                (* no progress has been made *)
+                stream_from (chunk_no + 1) offset
+              end else begin
+                let buffer = if (Int64.of_int this_chunk) = chunk_size
+                  then reusable_buffer
+                  else Bytes.make this_chunk '\000'
+                in
+                Unixext.really_read ifd buffer 0 this_chunk;
+                if not (Zerocheck.is_all_zeros (Bytes.unsafe_to_string buffer) this_chunk) || first_or_last then begin
+                  last_transmission_time := now;
+                  write_block ~__context filename buffer ofd this_chunk;
+                end;
+                made_progress __context progress (Int64.of_int this_chunk);
+                stream_from (chunk_no + 1) (Int64.add offset chunk_size);
+              end
+            end
+        in
+        stream_from 0 0L
+      | Some (path, exportname) ->
+          let actually_write_chunk (this_chunk_no: int) (this_chunk_size: int) =
+            let buffer = if this_chunk_size = Int64.to_int chunk_size
+              then reusable_buffer
+              else Bytes.make this_chunk_size '\000'
+            in
+            let filename = Printf.sprintf "%s/%08d" prefix this_chunk_no in
+            Unix.LargeFile.lseek ifd (Int64.mul (Int64.of_int this_chunk_no) chunk_size) Unix.SEEK_SET |> ignore;
+            Unixext.really_read ifd buffer 0 this_chunk_size;
+            last_transmission_time := Unix.gettimeofday ();
+            write_block ~__context filename buffer ofd this_chunk_size;
+            made_progress __context progress (Int64.of_int this_chunk_size);
+          in
+          let rec stream_from_offset (offset: int64) =
+            let remaining = Int64.sub size offset in
+            if remaining > 0L then begin
+              let this_chunk_size = min (Int64.to_int chunk_size) (Int64.to_int remaining) in
+              let this_chunk_no = Int64.div offset chunk_size in
+              let now = Unix.gettimeofday () in
+              let time_since_transmission = now -. !last_transmission_time in
+              if offset = 0L || remaining <= chunk_size || time_since_transmission > 5. then begin
+                actually_write_chunk (Int64.to_int this_chunk_no) this_chunk_size;
+                stream_from_offset (Int64.add offset (Int64.of_int this_chunk_size))
+              end else begin
+                let remaining = Int64.mul (Int64.div (Int64.sub remaining 1L) chunk_size) chunk_size in
+                (* Get sparseness for next 10GB or until the end rounded by last chunk, whichever is smaller *)
+                let sparseness_size = min max_sparseness_size remaining in
+                let output, _ = Forkhelpers.execute_command_get_output "/opt/xensource/libexec/get_nbd_extents.py" ["--path"; path; "--exportname"; exportname; "--offset"; Int64.to_string offset; "--length"; Int64.to_string sparseness_size] in
```

Comment:
Might want to put the path into Xapi_globs. There is also a way for xapi to check for presence of the script on startup if you list it in required resources.

Reply:
mmm, not sure I follow this, what is Xapi_globs?

Reply:
ocaml/xapi/xapi_globs.ml

Reply:
OK i will do that